### PR TITLE
Implement subscription using channels.

### PR DESCRIPTION
**Pull request by YoEight**
It only concerns regular subscriptions. A patch for persistent subscriptions will follow.

### CODE REVIEW

File: client/subscriptions.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+func (sub *Subscription) Close() error {
+	return sub.inner.CloseSend()
+}
+
+func (sub *Subscription) Recv() *SubscriptionEvent {
+	req := make(chan *SubscriptionEvent)
+
+	sub.channel <- req
```
Comment (pvanbuijtene) [inline client/subscriptions.go:133]: No idea how much it affects performance, but maybe it's an idea to look into using a single channel.

**Review by pvanbuijtene — APPROVED** [review]

File: client/subscriptions.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	client  *Client
+	id      string
+	inner   api.Streams_ReadClient
+	channel chan (chan *SubscriptionEvent)
+}
+
+func NewSubscription(client *Client, inner api.Streams_ReadClient, id string) *Subscription {
+	channel := make(chan (chan *SubscriptionEvent))
+
+	// It is not safe to consume a stream in different goroutines. This is why we only consume
```
Comment (pvanbuijtene) [inline client/subscriptions.go:35]: Is it the idea that a subscription can be used by multiple go routines ?
Would think it should be 1 goroutine per subscription 🤔

File: client/subscriptions.go (hunk 1 above)
Comment (YoEight) [reply client/subscriptions.go:133]: Then I don't khow to make that possible while keep using channels. An alternative implementation would be to use a mutex around the subscription stream.

File: client/subscriptions.go (hunk 2 above)
Comment (YoEight) [reply client/subscriptions.go:35]: Go gRPC docs say that sharing a read-stream among different goroutines is unsafe. It's safe for a write-stream though

File: client/subscriptions.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+			case *api.ReadResp_Checkpoint_:
+				{
+					checkpoint := result.GetCheckpoint()
+					position := position.Position{
+						Commit:  checkpoint.CommitPosition,
+						Prepare: checkpoint.PreparePosition,
+					}
+
+					req.channel <- &SubscriptionEvent{
+						CheckPointReached: &position,
```
Comment (hayley-jean) [inline client/subscriptions.go:96]: Now that checkpoints are handled differently, should we remove the `CheckpointReached` func on the `NewSubscriptionFilterOptions` as that callback does nothing now.

**Review by hayley-jean — CHANGES_REQUESTED** [review]
Looks good! One other thing that came up when testing is that when I call `Subscription.Close()` I only get notified of the subscription being dropped when the next event is handled.

It also looks like the subscription isn't disposed on the server, so the server will continue pushing events as they come in even after the subscription has been cancelled.

File: client/subscriptions.go (hunk 3 above)
Comment (YoEight) [reply client/subscriptions.go:96]: done

**Review by hayley-jean — APPROVED** [review]

**Review by pvanbuijtene — APPROVED** [review]

**Pull request merged**